staging/src/k8s.io/kubectl/pkg/cmd/delete/delete.go

```diff
@@ -291,20 +296,28 @@ func (o *DeleteOptions) RunDelete(f cmdutil.Factory) error {
 }
 
 func (o *DeleteOptions) DeleteResult(r *resource.Result) error {
-	found := 0
 	if o.IgnoreNotFound {
 		r = r.IgnoreErrors(errors.IsNotFound)
 	}
+
+	deletedInfos, err := r.Infos()
+	if len(deletedInfos) == 0 {
+		fmt.Fprintf(o.Out, "No resources found\n")
+		return err
+	}
+	if o.Confirm && !o.confirmation(deletedInfos) {
+		fmt.Fprintf(o.Out, "deletion is cancelled\n")
+		return err
+	}
+	var errs []error
+	if err != nil {
+		errs = append(errs, err)
+	}
+
 	warnClusterScope := o.WarnClusterScope
-	deletedInfos := []*resource.Info{}
 	uidMap := cmdwait.UIDMap{}
-	err := r.Visit(func(info *resource.Info, err error) error {
```

Comment:
I'd leave the visitor pattern here, as mentioned elsewhere we'd be invoking deletion and asking user interactively, that's the whole point, no?

Reply:
I tried to preserve visitor pattern in here but couldn't find a way to fetch all resources prior to delete operation.

Reply:
It definitely makes me nervous that the changes you had to make below are essentially replicating behaviour that was configured on the visitor IIUC. For example, it seems to me the code below assumes `ContinueOnError` was set, but that setting on the builder isn't actually tied to the behaviour anymore. Does this perhaps point to a missing visitor feature or visitor type that would provide the hook we need?

Reply:
Thanks for review @KnVerey. It makes me nervous too. Moreover, most of your questions above are all related to preserve the same functionality which already exists when `Visit` function is used.

What do you think about using a different builder(with same configurations) to list all resources(having an new builder `rpreview` like in here https://github.com/kubernetes/kubernetes/blob/3bf57e3ded6d69d4a7c03de272bb1f0f1c3ea8e4/staging/src/k8s.io/kubectl/pkg/cmd/delete/delete.go#L200), instead removing `Visit` functionality. That will assure that we are not breaking anything.

Reply:
Browsing around the code, it seems like the Builder is complex and monolithic, whereas there are already multiple Visitor types that can be plugged in for use with it. Indeed, I'm seeing that several Builder internals are implemented with decorator/helper Visitor types:

https://github.com/kubernetes/kubernetes/blob/920fb984eff204ea1371ccaf4c6191a6da41c70b/staging/src/k8s.io/cli-runtime/pkg/resource/builder.go#L1157-L1183

And the purposes of the decorator pattern here seems to be to give the 'helper' visitors a chance to inspect the set and potentially throw errors before the 'real' visitor is invoked. That sounds a lot like what we want? So maybe a preview / set confirmation decorator added by a confirmation option on the builder?

Reply:
@KnVerey  this seemed like a good idea and I tried to figure out can we use new visitor or not. However, visitor pattern does not cover in our case.

Because we need to traverse all list twice separately (one for preview and asking confirmation and the other one is for actual delete). But visitor only walks through one resource at a time and moves to another resource after completing walking through all visitors.

I used another builder to overcome that problem. There is no backward compatibility risk and works without any problem. What do you think?